server/raft_transport.go

```diff
@@ -101,43 +115,101 @@ func (t *rpcTransport) Stop(id proto.RaftNodeID) {
 	delete(t.servers, id)
 }
 
-// Send a message to the recipient specified in the request.
-func (t *rpcTransport) Send(req *multiraft.RaftMessageRequest) error {
-	// Convert internal to proto formats.
-	protoReq := &proto.RaftMessageRequest{GroupID: req.GroupID}
-	var err error
-	if protoReq.Msg, err = req.Message.Marshal(); err != nil {
-		return err
+func (t *rpcTransport) processQueue(nodeID proto.RaftNodeID) {
+	t.mu.Lock()
+	ch, ok := t.queues[nodeID]
+	t.mu.Unlock()
+	if !ok {
+		return
 	}
-
-	nodeID, _ := proto.DecodeRaftNodeID(proto.RaftNodeID(req.Message.To))
-	addr, err := t.gossip.GetNodeIDAddress(nodeID)
-	if err != nil {
-		return err
+	// Clean-up when the loop below shuts down.
+	defer func() {
+		t.mu.Lock()
+		delete(t.queues, nodeID)
+		t.mu.Unlock()
+	}()
+	var req *multiraft.RaftMessageRequest
+	protoReq := &proto.RaftMessageRequest{}
+	for {
+		select {
+		case <-t.rpcContext.Stopper.ShouldStop():
+			return
+		case <-time.After(raftIdleTimeout):
+			return
+		case req = <-ch:
+		}
+		if req == nil {
+			continue
+		}
+
+		// Convert to proto format.
+		protoReq.Reset()
+		protoReq.GroupID = req.GroupID
+		var err error
+		if protoReq.Msg, err = req.Message.Marshal(); err != nil {
+			log.Errorf("could not marshal message: %s", err)
+		}
+
+		nodeID, _ := proto.DecodeRaftNodeID(proto.RaftNodeID(req.Message.To))
+		addr, err := t.gossip.GetNodeIDAddress(nodeID)
+		if err != nil {
+			log.Errorf("could not get address for node %d: %s", nodeID, err)
+		}
+
+		deadline := time.After(raftMessageTimeout)
+		client := rpc.NewClient(addr, nil, t.rpcContext)
+		// TODO(tschottdorf) should let MultiRaft know that the node is gone;
+		// need a feedback mechanism for that.
+		// Potentially easiest is to arrange for the next call to Send() to
+		// fail appropriately.
+		select {
+		case <-t.rpcContext.Stopper.ShouldStop():
+			return
+		case <-client.Closed:
+			log.Errorf("raft client failed to connect")
+			continue
+		case <-deadline:
+			log.Errorf("call timed out after %s", raftMessageTimeout)
+			continue
+		case <-client.Ready:
+		}
+
+		call := client.Go(raftMessageName, protoReq, &proto.RaftMessageResponse{}, nil)
+		select {
+		case <-t.rpcContext.Stopper.ShouldStop():
+			return
+		case <-client.Closed:
+			log.Errorf("raft client failed to connect")
+			continue
+		case <-deadline:
+			log.Errorf("call timed out after %s", raftMessageTimeout)
+			continue
+		case <-call.Done:
```

Comment:
You're serializing raft message sending here. Is it desirable to send multiple raft messages to a node concurrently? If yes, I think we can restructure this just a little bit to pass in our own buffered `done` channel and then use a single select statement inside the loop. But perhaps that is better left for a TODO and a subsequent change.